### PR TITLE
Changing `Decoder` trait to be more composable. (#938)

### LINKED ISSUE #872: Add an `Sequence` object to the decoders
I wonder if it would be useful to have a sequence object for the decoders too.

It seems to me for example that if we build a tokenizer with a BPE model that defines a `end_of_word_suffix`, we will need to use the `BPEDecoder` decoder to replace the`end_of_word_suffix` and if we also used a ByteLevel pre-tokenization we will need the `ByteLevel` decoder to  realign the codes.

At the moment, it seems to me that we don't have a solution to choose a suitable decoder for such a tokenizer.

What do you think? :smile:

### PR DESCRIPTION
**Pull request by Narsil**
* Changing `Decoder` trait to be more composable.

### CODE REVIEW

**Comment by mishig25** [comment]
Got 2 questions:
1. What is the difference between #938 & #1008 and why was #938 reverted in the first place?
2. Do you need to create `tokenizers::decoders::Sequence`?

**Comment by Narsil** [comment]
> Got 2 questions:
> 
>     1. What is the difference between [Changing `Decoder` trait to be more composable. #938](https://github.com/huggingface/tokenizers/pull/938) & [Changing `Decoder` trait to be more composable. (#938) #1008](https://github.com/huggingface/tokenizers/pull/1008) and why was [Changing `Decoder` trait to be more composable. #938](https://github.com/huggingface/tokenizers/pull/938) reverted in the first place?
> 

This one is not breaking
>     2. Do you need to create `tokenizers::decoders::Sequence`?
Yes.

File: tokenizers/src/tokenizer/mod.rs [hunk 1]
```diff
@@ -119,9 +119,13 @@ impl dyn PostProcessor {
     }
 }
 
-/// A `Decoder` has the responsibility to merge the given `Vec<String>` in a `String`.
+/// A `Decoder` changes the raw tokens into its more readable form.
 pub trait Decoder {
-    fn decode(&self, tokens: Vec<String>) -> Result<String>;
+    fn decode(&self, tokens: Vec<String>) -> Result<String> {
+        let results = self.decode_chain(tokens)?;
+        Ok(results.join(""))
```
Comment (mishig25) [inline tokenizers/src/tokenizer/mod.rs:126]: will this line cause any problem for the SequenceDecoder? 
Specifically, one unlikekly edge-case I'm thinking is: one of the sequence decoders is another sequence decoder, and `Ok(results.join(""))` will cause an informaition loss?

as a solution, we can implement `fn decode` for Sequence Decoder?

Comment (Narsil) [reply tokenizers/src/tokenizer/mod.rs:126]: It's OK I think.

The parent `Sequence` would call the child `Sequence.decode_chain` NOT the `decode` function, so we're good, no ?

I could add a test to make sure.

Comment (mishig25) [reply tokenizers/src/tokenizer/mod.rs:126]: ooh I see, so it is no problem in that case

**Review by mishig25 — APPROVED** [review]
lgtm!

**Pull request merged**